File: dana/dana/test/unit.py
#!/usr/bin/env python
#
# Copyright (c) 2006-2007 Nicolas Rougier.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
# 
# $Id$

import unittest
import dana.core as core


class UnitTests (unittest.TestCase):
    def setUp (self):
        self.unit = core.Unit()
        
    def testPotential (self):
        """ Check unit default potential is 0.0 """
        self.assertEqual (self.unit.potential, 0.0)

    def testSpec (self):
        """ Check unit default spec is None """
        self.assertEqual (self.unit.spec, None)

    def testPosition (self):
        """ Check unit default position is (-1,-1) """
        self.assertEqual (self.unit.position, (-1,-1))


    def testSetPosition(self):
        """ Check unit position is read-only """
        self.assertRaises (AttributeError, setattr, self.unit,
                           'position', (0,0))
    def testWeightsNoArg(self):
        """ Check unit weights cannot be get without argument """
        self.assertRaises (TypeError, self.unit.weights)

    def testWeightsNoneArg (self):
        """ Check unit weigths cannot be get with None argument """
        self.assertRaises (AssertionError, self.unit.weights, None)

    def testWeightsGarbage (self):
        """ Check unit weights cannot be get with argument of wrong type """
        self.assertRaises (TypeError, self.unit.weights, "ee")

    def testComputeDP (self):
        """ Check unit potential computation """
        self.assertEqual (self.unit.compute_dp(), 0.0)

    def testComputeDW (self):
        """ Check unit weights computation """
        self.assertEqual (self.unit.compute_dw(), 0.0)

    def testConnect (self):
        """ Check unit connect function """    
        other = core.Unit()
        self.unit.connect (other, 1.0)

# Test suite
suite = unittest.TestLoader().loadTestsFromTestCase(UnitTests)

if __name__ == "__main__":
    unittest.main()

File: dana.life/dana/life/unit.cc
//
// Copyright (C) 2006 Nicolas Rougier
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
//
// $Id$

#include "dana/core/map.h"
#include "dana/core/layer.h"
#include "dana/core/link.h"
#include "unit.h"

using namespace boost::python::numeric;
using namespace dana;
using namespace dana::life;

// ============================================================================
//  constructor
// ============================================================================
Unit::Unit(void) : core::Unit()
{
}

// ============================================================================
//  destructor
// ============================================================================
Unit::~Unit(void)
{}

// ============================================================================
//  computes potential
// ============================================================================
float
Unit::compute_dp (void)
{
    float n= 0;
    for (unsigned int i=0; i<laterals.size(); i++)
        n += laterals[i]->source->potential;
    
    if ((potential == 0) && (n == 3))
        _potential = 1.0;
    else if ((n < 2) || (n > 3))
        _potential = 0.0f;

    potential = _potential;

	return 0;
}

// ============================================================================
//  computes weights
// ============================================================================
float
Unit::compute_dw (void)
{
    potential = _potential;
	return 0;
}


// ============================================================================
//    Boost wrapping code
// ============================================================================
void
Unit::boost (void) {

    using namespace boost::python;
    register_ptr_to_python< boost::shared_ptr<Unit> >();
    
    class_<Unit, bases<core::Unit> >("Unit",
    "======================================================================\n"
    "\n"
    "A unit is a potential that is computed on the basis of some external\n"
    "sources that feed the unit. Those sources can be anything as long as\n"
    "they have some potential.\n"
    "\n"
    "Attributes:\n"
    "-----------\n"
    "   potential : unit potential (float)\n"
    "   position  : unit position within layer (tuple, read only)\n"
    "\n"
    "======================================================================\n",
        init<>(
        "__init__ () -- initialize unit\n")
        )
        ;
}
